docs/standards/rest.md

```diff
@@ -201,6 +201,128 @@ The `version` URL query parameter, `?version=version_string` is reserved for sel
+
+To expand related resources in a one-to-many relationship, the outermost resource may be a collection. Then the "to-one" related resource may be expanded within each item.
+
+#### Examples
+
+To illustrate this, consider a resource API of `books`, `series` and `authors`, where one author is related to many books, and books may be part of a series. (A real books API would have to deal with books having multiple authors but we'll simplify things here for sake of example.)
+
+The author and series relationships for each book in a resource collection response may be expanded, as these are "to-one" with each book:
+
+```
+GET /books?expand=author,series
+{
+  "data": [{
+    "id": "cd4f014a-0aff-4fe9-8cd1-0c66550e2461",
+    "type": "book",
+    "attributes": {
+      "title": "Consider Phlebas",
+      "isbn": "031600538X"
+    },
+    "relationships": {
+      "author": {
+        "data": {
+          "id": "5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea",
+          "type": "author",
+          "attributes": {
+            "name": "Iain M. Banks"
+          }
+        }
+      },
+      "series": {
+        "data": {
+          "id": "fc35a72f-bb43-47c9-910b-efd495da4fc9",
+          "type": "series",
+          "attributes": {
+            "name": "The Culture"
+          }
+        }
+      }
+    }
+  },{
+    "id": "73fea654-e7f1-40b8-819c-708cabaea537",
+    "type": "book",
+    "attributes": {
+      "title": "The Player of Games",
+      "isbn": "0708883095"
+    },
+    "relationships": {
+      "author": {
+        "data": {
+          "id": "5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea",
+          "type": "author",
+          "attributes": {
+            "name": "Iain M. Banks"
+          }
+        }
+      },
+      "series": {
+        "data": {
+          "id": "fc35a72f-bb43-47c9-910b-efd495da4fc9",
+          "type": "series",
+          "attributes": {
+            "name": "The Culture"
+          }
+        }
+      }
+    }
+  }, /* ... */],
+  "links": { /* pagination links */ }
+}
+```
+
+Relationship links to a resource collection may still be used to express and navigate a "to-many" relation -- these just can't be expanded inline:
+
+```
+GET /authors/5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea
+{
+  "data": {
+    "id": "5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea",
+    "type": "author",
+    "attributes": {
+      "name": "Iain M. Banks"
+    }
+    "relationships": {
+      "books": {
+        "links": {
+          "related": "/books?author_id=5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea"
+        }
+      }
+    }
+  }
+}
+```
+
+```
+GET /authors/5c4b6ec5-9ca0-4b14-8ab1-23ee26684cea?expand=books
+
+HTTP 400 Bad Request
+Content-Type: application/vnd.api+json
+{
+  "errors": [{ /* can't do this */ }]
```

Comment:
Do we want to error on this? Or just not expand things that we don't support expanding? I get "explicit better than implicit" but I can also see this making things more complicated on the consumer side where you just want to expand everything that can be expanded.

But perhaps that sort of aggro API consumption isn't a thing we should support, as I think about it?

Reply:
I think we'd 400 here specifically because you'd be specifying an invalid enum value for `?expand=` if you tried to do this. So you'd basically be dealing with a validation error at this point.

Sweater Comb rules would also prevent it from happening in API design in the first place. In a followup, we can add a rule that does not allow a relationship object's `.data` property to be an array.

Reply:
"Expand all" is interesting. I think it'd be fine as an optional shorthand, but I'd like to gauge demand before adding it to the spec.

If you're building UI components for expanded resources, I think you're probably having to integrate specifically with each expansion, so it might be fine to expect those to be enumerated.

I also didn't want to require all relationships to support expansion. In some cases it might be expensive or complicated to expand, or to do so with consistency guarantees, without necessary supporting infrastructure. So this gives the resource owner some breathing room to link first, then support inlining when it's needed, and release to beta / GA when it can be made performant / reliable.

Reply:
This seems pretty limiting for an obvious use case, the author-books example is illustrative.  Our users will want to see projects->targets, groups->orgs, orgs->users, etc.  Many of the obvious use cases require one->many expansion so not supporting that out of the gate feels not great.

Reply:
I think it is supported, you just need to "flip it around" so that you're expanding items in a paginated collection.

So in your examples, I think you could:

- `GET /orgs/{org_id}/projects?expand=target` to get a collection of projects with their target expanded in each item
- `GET /orgs?expand=group` to get a collection of orgs with their group expanded
- `GET /users?expand=org` to get a collection of users with their org expanded

The only use case this doesn't support is a case where you want to deal with multiple collections simultaneously, such as an N:M relationship. This becomes really difficult to express pagination over however.